willow/image.py

```diff
@@ -124,7 +130,60 @@ def save(self, image_format, output):
             raise ValueError("Unknown image format: %s" % image_format)
 
         operation_name = "save_as_" + image_format
-        return getattr(self, operation_name)(output)
+        return getattr(self, operation_name)(output, apply_optimizers=apply_optimizers)
+
+    def optimize(self, image_file, image_format: str):
+        """
+        Runs all available optimizers for the given image format on the given image file.
+
+        If the passed image file is a SpooledTemporaryFile or just bytes, we are converting it to a
+        NamedTemporaryFile to guarantee we can access the file so the optimizers to work on it.
+        If we get a string, we assume it's a path to a file, and will tyr use it directly.
+        """
+        optimizers = registry.get_optimizers_for_format(image_format)
+        if not optimizers:
+            return
+
+        has_named_file = False
+        if isinstance(image_file, SpooledTemporaryFile):
```

Comment:
The logic for this method is: create a NamedTemporaryFile if needed, optimize in place, and replace the passed `image_file`. If `image_file` is a string, we assume a path and try to optimize that directly

note: this is probably the most controversial bit.
Latest Wagtail passes on a `SpooledTemporaryFile` as of https://github.com/wagtail/wagtail/pull/10284.

Given willow is used in non-Wagtail contexts, I wanted to support: `BytesIO` (which is what Wagtail passed prior to https://github.com/wagtail/wagtail/pull/10284), `NamedTemporaryFile`, + string and bytes

The reason for string is that we want to support file paths (https://github.com/wagtail/Willow/issues/108). I am unsure about bytes, though.